### DISCUSSION TITLE
Interface constructors

### DISCUSSION THREAD

**Discussion by AraHaan** [dotnet/csharplang#5119] [Language Ideas]
Suppose for example you are writing an interface, of which you want the users who derive from said interface to not worry about:

- needing to "implement" auto properties within the interface (would need an option or keyword to mark that the non-static properties within the interface as not required to "implement" on derived types which would need to be a separate issue / proposal)
- needing to implement a common constructor inside of the interface instead of the derived classes / structs / records

Where basically that interface constructor is then in charge of setting the properties defined within the interface to values passed into it. For example assume let's look into some code using Remora.Discord.Commands:

```cs
using JetBrains.Annotations;
using Remora.Discord.API.Abstractions.Rest;
using Remora.Discord.Commands.Contexts;

namespace MyBot.Commands;

[PublicAPI]
public interface ICommands
{
    public ICommandContext Context { get; set; }
    public IDiscordRestInteractionAPI DiscordRestInteractionApi { get; set; }

    public ICommands(ICommandContext context, IDiscordRestInteractionAPI discordRestInteractionApi)
    {
        this.Context = context;
        this.DiscordRestInteractionApi = discordRestInteractionApi;
    }
}
```

Where basically in this example the user intends to use this interface to implement for them the constructor to all of their types that derive from the way Remora creates commands for Discord, all of which basically do the same exact thing so they felt like they should extract that constructor to an interface instead, of which the arguments are then passed in to their type using DependencyInjection.

I am sure there are other use cases as well where interfaces being used to implement common constructors in types could also help reduce code duplications as well too (especially for common constructors).

**Comment by huoyaoyuan** [comment]
> needing to "implement" auto properties within the interface (would need an option or keyword to mark that the non-static properties within the interface as not required to "implement" on derived types which would need to be a separate issue / proposal)

Interface cannot have fields, so auto properties won't work anyway.
Supporting fields is a huge problem in type system and probably won't happen.

**Comment by budgetdevv** [comment]
Hmmm, how about something like:

```csharp
public interface ICommands
{
    public ICommands(ICommandContext context, IDiscordRestInteractionAPI discordRestInteractionApi);
}
```

Then you could do...

```csharp
public void InstantiateCommand<T>(ICommandContext context, IDiscordRestInteractionAPI discordRestInteractionApi) where T: ICommands
{
  var Cmd = new T(context, discordRestInteractionApi);
}
```

I've always dreamt of this

**Comment by budgetdevv** [comment]
I've also thought of something like

```cs
public void InstantiateCommand<T>() where T: new(ICommandContext, IDiscordRestInteractionAPI)
{
}
```
...but that's kinda irrelevant to this thread, I suppose

**Reply by TonyValenti** [reply]
I think the upcoming static interface methods might let you do what you want.￼￼

**Reply by budgetdevv** [reply]
The static abstracts? Those could work, but I'd love to retain the ability to set read-only fields

**Reply by TonyValenti** [reply]
Something else you may consider is using records instead of classes. I don't know your use case, but where I work, we've pretty much outlawed using classes to store data and we've gained lots of benefits by doing so - the read only and immutability benefits are just one.

**Reply by budgetdevv** [reply]
I do not see how using records over classes solve the issue of not being able to new() generics with a subset of parameters
